### ISSUE TITLE
Add custom features to NER

### ISSUE THREAD

**Issue by jamesmf** [RasaHQ/rasa#4214]
**Description of Problem**:
Right now custom entities can only use `pos` features from `spacy` and a handful of simple features. This seems to be in contrast to the flexibility and power of the other pipeline components which can take advantage of any combination of built-in and custom `featurizers`. Ideally, there would be a way to pass `ner_features` to the `CRFEntityExtractor`. In particular, this would let you train NER that used word/token vectors straight from spacy (or other pretrained models)

**Overview of the Solution**:
- `CRFEntityExtractor` needs to additionally check for `ner_features` on the message and add them to the feature dict it passes to `sklearn_crfsuite`.
- There need to be NER featurizer classes added

**Examples** (if relevant):
The skeleton of this (both adding a `spacy`-based featurizer and making `CRFEntityExtractor` use `ner_features`) is implemented in this PR
https://github.com/RasaHQ/rasa/pull/4187
Please let me know if this looks like a useful feature and if this PR is heading in the right direction.

Still necessary:
- Add tests
- Extend `Featurizer` to also have `_combine_with_existing_ner_features` 
- Validate that having default spacy tokens noticeably improves NER for a sample task
- Make `spacy` only optionally add to `ner_features`
- Replace the hard-coded lambda functions in `CRFEntityExtractor` with a simple `Featurizer`

**Definition of Done**:
<!-- What needs to be there to consider this feature as done?
- [ ] Tests are added
- [ ] Feature described the docs
- [ ] Feature mentioned in the changelog

**Comment by sara-tagger** [comment]
Thanks for submitting this feature request 🚀@MetcalfeTom will get back to you about it soon!✨
###### Caw! 🐧

**Comment by amn41** [comment]
thanks @jamesmf ! I really like the general idea + its flexibility, very cool proposal. Regarding your "still necessary" list, I agree an important step is showing that this improves performance on 1-2 example datasets.

**Comment by jamesmf** [comment]
Here's the experiment I ran.

Config.yml looks like this:
```
pipeline:
  - name: "SpacyNLP"
  - name: "SpacyTokenizer"
  - name: "SpacyFeaturizer"
  - name: "CRFEntityExtractor"
    features: [ ["low", "title", "upper", "ner_features"],
      [
          "bias",
          "ner_features",
      ],
     ["low", "title", "upper", "ner_features"]]

```

Config_orig.yml only uses the default features before adding `ner_features`
```
pipeline:
  - name: "SpacyNLP"
  - name: "SpacyTokenizer"
  - name: "SpacyFeaturizer"
  - name: "CRFEntityExtractor"
    features: [ ["low", "title", "upper"],
      [
          "bias",
          "low",
          "prefix5",
          "prefix2",
          "suffix5",
          "suffix3",
          "suffix2",
          "upper",
          "title",
          "digit",
          "pattern",
      ],
     ["low", "title", "upper"]]
```

I then ran a `rasa test nlu --config config.yml config_orig.yml --nlu data/nlu.md  --runs 5 --percentages 25 50` inside `restaurantbot`.

These are the (mean) results from each `CRFEntityExtractor.json` in the result.
```
{
  "25%": {
    "micro avg": {
      "config": 0.9839785565918215,
      "config_orig": 0.979184211632882
    },
    "macro avg": {
      "config": 0.9537193190641351,
      "config_orig": 0.896015038647866
    },
    "weighted avg": {
      "config": 0.9839827425204907,
      "config_orig": 0.9786516417647068
    }
  },
  "50%": {
    "micro avg": {
      "config": 0.9785380738557364,
      "config_orig": 0.9744053746311708
    },
    "macro avg": {
      "config": 0.9454309319154636,
      "config_orig": 0.8360626493692491
    },
    "weighted avg": {
      "config": 0.978586347904418,
      "config_orig": 0.9735714854886236
    }
  }
}

```

You can see the new (`config`) `CRFEntityExtractor` that uses primarily `ner_features` from a `en_core_web_md` `spacy` model slightly outperforms the original (`config_orig`) across all metrics at both holdout values. I did not test adding `ner_features` on top of ALL the existing features, though.

**Comment by Zylatis** [comment]
I've been considering a similar thing and one particular feature I am hoping to implement, and would be great to see in this implementation, is 'distance to keyword'. For example, I am trying to extract destination and origin addresses and a feature which may be important is the distance of a given token/word to a 'from' or 'to' word (from a list) in that sentence.

In terms of this PR would the idea be that i create this as a customer featuriser which can plug into this new config?

**Comment by jamesmf** [comment]
@Zylatis yes, if you wanted to add that as a component compatible with this, I would imagine you'd create something like this:

```
class KeywordDistanceFeaturizer(Featurizer):
    ...
    def process(self, message):
         # compute an np array of shape (num_tokens, num_features) where each (i, j) index might be the ith token's distance from the jth keyword
         keyword_features = self.get_keyword_distances(message.get("tokens", []))
         self._combine_with_existing_ner_features(message, keyword_features):
```

**Comment by Zylatis** [comment]
Cool, would definitely be handy. I'm definitely keen to do as much custom feature engineering with this CRF as possible, so if you'd like help on this PR let me know (not an expert by any means but i'd like to contribute if i can). @jamesmf
